std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,690 @@
+                    lock.lock();
+                }
+                else
+                {
+                    tmp.bytesUsed += n;
+                }
+
+                // Most likely this node has memory for more allocations
+                // Move it to the front
+                moveToFront(tmp);
+
+                static if (isShared)
+                {
+                    tmp.keepAlive--;
+                    if (next) next.keepAlive--;
+                }
+
+                return result;
+            }
+
+            // This node can now be removed if necessary
+            static if (isShared)
+            {
+                lock.lock();
+                tmp.keepAlive--;
+                if (next) next.keepAlive--;
+            }
+
+            if (!next)
+                break;
+
+            tmp = next;
+            next = tmp.next;
+
+            // If there are too many nodes, free memory by removing empty nodes
+            static if (isShared)
+            {
+                if (atomicLoad(numNodes) > maxNodes &&
+                    atomicLoad(tmp.bytesUsed) == 0 &&
+                    tmp.keepAlive == 0)
+                {
+                    removeNode(tmp);
+                }
+            }
+            else
+            {
+                if (numNodes > maxNodes && tmp.bytesUsed == 0)
+                {
+                    removeNode(tmp);
+                }
+            }
+
+            tmp = next;
+        }
+
+        // Cannot create new AlignedBlockNode. Most likely the ParentAllocator ran out of resources
+        if (!insertNewNode())
+            return null;
+
+        tmp = cast(typeof(tmp)) root;
+        void[] result = tmp.bAlloc.allocate(n);
+
+        static if (isShared)
+        {
+            atomicOp!"+="(root.bytesUsed, result.length);
+        }
+        else
+        {
+            root.bytesUsed += result.length;
+        }
+
+        return result;
+    }
+
+    // goodAllocSize should not use state
+    size_t goodAllocSize(const size_t n)
+    {
+        Allocator a = null;
+        return a.goodAllocSize(n);
+    }
+}
+
+/**
+`AlignedBlockList` represents a wrapper around a chain of allocators, allowing for fast deallocations
+and preserving a low degree of fragmentation.
+The allocator holds internally a doubly linked list of `Allocator` objects, which will serve allocations
+in a most-recently-used fashion. Most recent allocators used for `allocate` calls, will be
+moved to the front of the list.
+
+Although allocations are in theory served in linear searching time, `deallocate` calls take
+$(BIGOH 1) time, by using aligned allocations. `ParentAllocator` must implement `alignedAllocate`
+and it must be able to allocate `theAlignment` bytes at the same alignment. Each aligned allocation
+done by `ParentAllocator` will contain metadata for an `Allocator`, followed by its payload.
+
+Params:
+    Allocator = the allocator which is used to manage each node
```

Comment:
The documentation does an insufficient job of explaining what needs to be true of the Allocator!

Reply:
Yes, you're right, I haven't explained here too much about the `Allocator`.
`Allocator` can be anything, as long as it has a constructor which receives `ubyte[]` and its `ParentAllocator` is the `NullAllocator`.

For the moment, these are the only constraints which come to mind.
Thanks!